### DISCUSSION TITLE
How to propagate the notification for a collection of item that implements INotifyPropetyChanged?

### DISCUSSION THREAD

**Discussion by Snailya** [reactivemarbles/DynamicData#890] [Q&A] [answered]
I have a class that implements INotifyPropertyChange. Then i need a correct type of collection to holds these class items.
Whenver the property of the item changed, I need to build the collection into a tree structure.
Before, i used to use SourceCache and update the source cache in a service using AddOrUpdate. But now I want the item to handle its update by itself, not a service.

**Comment by JakenVeina** [comment]
> Whenver the property of the item changed, I need to build the collection into a tree structure.

Not really sure what you mean by this. Are you saying you want to derive a tree structure, based on properties of the items? Like each item has its `ParentId` as a notifying property?

**Reply by Snailya** [reply]
I'm creating an addin for visio which displays the shapes in the page as a tree structure. 
To make the tree item properties updates synchronously with the shape, i maked the shape model in my code subscribe to the event in visio. 
When initializing my tree, I firstly created a ObservableCollection. 
Then convet it to a sourcecache of ShpeViewModel, which is just a property copy with no event subscrption. 
Then to a tree structure using TransformToTree(). 
It readlly strange that, if i change the property of a shape which is not the root item, it works fun. But if i change the property of a root shape, it fails to reflect the change on UI and sub shape stop works etiher.
I don't know how to find out the bug, so I create a small WPF app, it works well.
Here is my code, maybe i'm doing the wrong way.
ShapeModel.cs
```
public abstract class ShapeModel: ReactiveObject, IDisposable {
    private readonly CompositeDisposable _cleanUp = new();
    protected readonly Visio.Shape Source;
    private int _parentId;
    private string _description = string.Empty;

   protected ShapeModel(Visio.Shape source)
    {
        Source = shape;
        Id = shape.ID;

        Observable
            .FromEvent<EShape_CellChangedEventHandler, Cell>(
                handler => Source.CellChanged += handler,
                handler => Source.CellChanged -= handler)
            .Subscribe(OnCellChanged)
            .DisposeWith(_cleanUp);

        // observable on relationship change
        Observable
            .FromEvent<EShape_FormulaChangedEventHandler, Cell>(
                handler => Source.FormulaChanged += handler,
                handler => Source.FormulaChanged -= handler)
            .Where(cell => cell.Name == "Relationships")
            .Subscribe(_ => { OnRelationshipsChanged(); })
            .DisposeWith(_cleanUp);
    }

    public int Id { get; }
    public int ParentId
    {
        get => _parentId;
        protected set => this.RaiseAndSetIfChanged(ref _parentId, value);
    }
    public string Description
    {
        get => _description;
        protected set => this.RaiseAndSetIfChanged(ref _description, value);
    }
}
```
ShapeViewModel.cs is a reduced version of ShapeModel.cs without the Subscription.
In my PageViewModel
```
        var sourceCache = new SourceCache<ShapeViewModel, int>(t => t.Id);
        service.Elements
            .ToObservableChangeSet(t => t.Id)
            .AutoRefresh()
            .ObserveOn(RxApp.MainThreadScheduler)
            .Subscribe(changes =>
            {
                sourceCache.Edit(updater =>
                {
                    foreach (var change in changes)
                        switch (change.Reason)
                        {
                            case ChangeReason.Remove:
                                updater.Remove(change.Current.Id);
                                break;
                            case ChangeReason.Add:
                            case ChangeReason.Update:
                            case ChangeReason.Refresh:
                            case ChangeReason.Moved:
                            default:
                                var item = new ShapeViewModel(change.Current);
                                updater.AddOrUpdate(item);
                                break;
                        }
                });
            })
            .DisposeWith(d);

        sourceCache
            .Connect()
            .TransformToTree(x => x.ParentId)
            .Transform(node => new BOMTreeItemViewModel(node))
            .ObserveOn(RxApp.MainThreadScheduler)
            .Bind(out _bomTree)
            .DisposeMany()
            .Subscribe()
            .DisposeWith(d);
```

**Comment by JakenVeina** [comment, marked as answer]
I'm not familiar with the details of how the tree transforms work, myself, but everything you've described laid out sounds and looks correct, logically, I don't think you're missing anything fundamental about proper usage of DD.

What I'm failing to see here is the connection between `ShapeModel` and `ShapeViewModel`. Or are you saying that `ShapeViewModel` is from your smaller demo app, and that version works, but it's not working with `ShapeModel` in the full app? I'd also want to see what `BOMTreeItemViewModel` has in it.

If it were me, my next step would be to toss `.Do(x => {})` into the stream, both before `.TransformToTree()` and after `.Transform()` and set breakpoints in them, in order to inspect exactly what changes are being published when you do whatever action it is that triggers a change to a root item. I would think you should see a `Replace` change in both cases `.TransformToTree()`. If you get nothing, the problem has gotta be upstream from where you populate `sourceCache`. If you see the first `Replace` but not the second one, either `BOMTreeItemViewModel` is somehow swallowing the changes, or there may be a bug within `.TransformToTree()` or `.Transform()`. If you see both `Replace` changes, start looking at your View layer, it may be that your `{Binding}`s are incorrect there, in some way that affects only root items.

**Reply by Snailya** [reply]
Thank you. Having struggle with it for hours, I think there is a high probability that something wrong with my custom control template.

**Reply by JakenVeina** [reply]
Happy to take a look at it.
